lib/prime.rb

```diff
@@ -32,15 +32,48 @@ def prime_division(generator = Prime::Generator23.new)
 
   # Returns true if +self+ is a prime number, else returns false.
   def prime?
-    return self >= 2 if self <= 3
-    return true if self == 5
-    return false unless 30.gcd(self) == 1
-    (7..Integer.sqrt(self)).step(30) do |p|
-      return false if
-        self%(p)    == 0 || self%(p+4)  == 0 || self%(p+6)  == 0 || self%(p+10) == 0 ||
-        self%(p+12) == 0 || self%(p+16) == 0 || self%(p+22) == 0 || self%(p+24) == 0
-    end
-    true
+    # https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test
+    # https://arxiv.org/pdf/1509.00864.pdf
+    #
+    # if n < 3,317,044,064,679,887,385,961,981,
+    # it is enough to test a = 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, and 41.
+    # and the result will be deterministic.
+    #
+
+    if self > 3317044064679887385961981
+      raise ArgumentError, ".prime? is only valid for values less than 3,317,044,064,679,887,385,961,981"
+    end
+    
+    return false if self < 2
+    return true  if self < 4
+    return false if self.even?
+
+    r = 0
```

Comment:
I believe you could start ``r`` with ``-1`` to later replace ``(r-1).times do ...`` with ``r.times do ...``.

Reply:
r is the number of times 2 divides self, so since its odd, r=1 and d = (self -1)/2....

Reply:
Why subtract one from self in ``d = (self - 1) / 2?``
The number is odd, 0bxxxx1, and you will divide by two, shifting right one position and removing that last one bit.
I believe the result is the same: (0bxxxx1 - 1) / 2 == 0bxxxx1 / 2 == 0b0xxxx

Reply:
The first step in M-R is
`write n as 2r·d + 1 with d odd (by factoring out powers of 2 from n − 1)`

So I'm looking for `2r * d  = (n-1)`

but if we know n is odd, we know it'll have to do atleast one loop iteration as n-1 is even.

So that shortcuts down to r=1 and d=(n-1)/2

Reply:
I still do not see the difference between ``d=(self-1)/2`` and ``d=self/2`` for odd ``self``.
If you think this is the right way to match the algorithm I am fine with it.

You will probably have to add a benchmark to show how much it improves over the current **prime?** before the PR is accepted.

Reply:
Oh that's right, / 2 is correct.

Reply:
I'll do a benchmark as well.